extra_geom/motors.py

```diff
@@ -0,0 +1,214 @@
+import numpy as np
+
+
+def read_motors_from_geom(text):
+    """Reads the motor positions from the text format."""
+    if isinstance(text, str):
+        import io
+        file = io.StringIO(text)
+    else:
+        file = text
+
+    meta = {
+        line[0]: line[2]
+        for line in (
+            line[7:].partition('=')
+            for line in file if line.startswith(";XGEOM ")
+        )
+    }
+    try:
+        n_groups, n_motors = (int(n) for n in meta["MOTORS"].split(','))
+    except KeyError:
+        raise ValueError("MOTORS record is not found")
+    except ValueError:
+        raise ValueError(
+            "Invalid MOTORS format, expected two comma-separated integers")
+
+    positions = []
+    for q in range(n_groups):
+        try:
+            key = f"MOTOR_Q{q+1}"
+            pos = [float(n) for n in meta[key].split(',')]
+        except KeyError:
+            raise ValueError(key + " record is not found")
+        except ValueError:
+            raise ValueError(
+                f"Invalid {key} format, expected {n_motors} "
+                "comma-separated floats")
+        if len(pos) != n_motors:
+            raise ValueError(
+                f"Wrong length of {key}, expected {n_motors} floats")
+
+        positions.append(pos)
+
+    return np.array(positions)
+
+
+class BaseMotorTracker:
+    """Detector motor tracker updates geometry according to motor positions.
+    """
+    def __init__(self, ref_geom):
```

Comment:
```suggestion
    def __init__(self, ref_geom, ref_motor_positions=None):
```

I think by adding an optional parameter here we can get rid of both `with_reference_positions()` and `set_reference_positions()`. If you want to use new reference positions, you'd just construct a new object:

```python
motor_tracker = AGIPD_1MMotors(motor_tracker, new_ref_positions)
```

Reply:
With two constructors, I'm trying to cover three cases:
1. The tracker without reference positions at all (default constructor). 
2. The tracker with explicitly given reference positions (the second constructor with parameter).
3. The tracker with reference positions which taken from reference geometry instance (the second constructor with default parameter).

In the first case, it's impossible to get geometry by absolute motor positions  (`geom_at` raises an exception)

In the third case, the reference positions must be in the geometry instance. Otherwise an exception is raised.

This is designed to implement the explicit behavior in every case, as your mentioned in the last comment in the previous time. And I like an idea.

In addition, `geom_at` creates the new geometry instance with updated reference position. If the instance is saved later, the reference positions go to the file as well.

In opposite, `move_geom_by` creates the new geometry instance without reference positions.

With one constructor and one parameter with default value is not possible to cover all three cases. It requires one more special value for the parameter or one more parameter. The second constructor looks more natural for me.

`set_reference_position` isolates the code which explicitly set the reference position in the second constructor. In principle, we can get rid of it. But why do not expose if have it anyway?

Reply:
I see, but I wonder if that's confusing if you have a geometry object with motor positions attached and you pass it to `motor_tracker = AGIPD_1MMotors(geom_with_motors)` - this will entirely ignore the motor positions. So I wonder about distinguishing cases 1 & 3 just by whether the geometry has motor information or not.

With regard to my comment from the last round, I agree that if you don't have reference positions, it shouldn't be possible to call `motors.geom_at()`. And I see there's a separate check for that. I'm not sure this needs to be separately checked in a constructor, but I don't feel strongly about this.

Reply:
Hmm... Then we don't have an option to discard reference motor position, but we free to ignore the call which is uses them. Ok, let's try to go with it.

Reply:
I've pushed changes